prow/cmd/exporter/collector_test.go

```diff
@@ -79,6 +88,222 @@ func assertEqual(t *testing.T, actual, expected interface{}) {
 	}
 }
 
+type falkLister struct {
+}
+
+func (l falkLister) List(selector labels.Selector) ([]*prowapi.ProwJob, error) {
+	return []*prowapi.ProwJob{
+		{
+			Spec: prowapi.ProwJobSpec{
+				Agent: prowapi.KubernetesAgent,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "default",
+				Name:      "7785d7a6-e601-11e9-8512-da8015665453",
+				Labels: map[string]string{
+					"created-by-prow":          "true",
+					"event-GUID":               "770bab40-e601-11e9-8e50-08c45d902b6f",
+					"preset-bazel-scratch-dir": "true",
+					"preset-service-account":   "true",
+					"prow.k8s.io/job":          "pull-test-infra-bazel",
+					"prow.k8s.io/refs.org":     "kubernetes",
+					"prow.k8s.io/refs.pull":    "14543",
+					"prow.k8s.io/refs.repo":    "test-infra",
+					"prow.k8s.io/type":         "presubmit",
+				},
+				Annotations: map[string]string{
+					"prow.k8s.io/job":            "pull-test-infra-bazel",
+					"testgrid-create-test-group": "true",
+				},
+			},
+		},
+		{
+			Spec: prowapi.ProwJobSpec{
+				Agent: prowapi.KubernetesAgent,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "default",
+				Name:      "e44f91e5-e604-11e9-99c1-0a58ac10f9a6",
+				Labels: map[string]string{
+					"created-by-prow":       "true",
+					"event-GUID":            "e4216820-e604-11e9-8cf0-295472589b4f",
+					"prow.k8s.io/job":       "branch-ci-openshift-release-master-config-updates",
+					"prow.k8s.io/refs.org":  "openshift",
+					"prow.k8s.io/refs.repo": "release",
+					"prow.k8s.io/type":      "postsubmit",
+				},
+				Annotations: map[string]string{
+					"prow.k8s.io/job": "branch-ci-openshift-release-master-config-updates",
+				},
+			},
+		},
+	}, nil
+}
+
+func TestProwJobCollector(t *testing.T) {
+
+	pjc := prowJobCollector{
+		lister: falkLister{},
+	}
+	c := make(chan prometheus.Metric)
+	go pjc.Collect(c)
+
+	now := time.Now()
+	counter := 0
+	var labelMetric1, annotationMetric1, labelMetric2, annotationMetric2 prometheus.Metric
+	waiting := true
+
+	for {
+		if !waiting {
+			break
+		}
+		select {
+		case msg := <-c:
+			counter++
+			logrus.WithField("counter", counter).Infof("received a metric")
+			switch counter {
+			case 1:
+				labelMetric1 = msg
+			case 2:
+				annotationMetric1 = msg
+			case 3:
+				labelMetric2 = msg
+			case 4:
+				annotationMetric2 = msg
+			}
+			if counter > 4 {
+				logrus.WithField("counter", counter).Errorf("should not have received more than 4 metrics, existing ...")
+				waiting = false
+			}
+		default:
+			if time.Since(now) > time.Second {
+				logrus.WithField("counter", counter).Errorf("timeout, existing ...")
```

Comment:
This should be a `Fatal` error

Reply:
I think this should even be an INFO since `timeout` is a valid existing from the loop (the other one is `counter>4`.

Reply:
Done.